emails/tests/views_tests.py

```diff
@@ -137,31 +139,110 @@ def create_email_from_notification(
     return email.as_bytes()
 
 
-def assert_email_equals(output_email: str, name: str) -> None:
+def create_notification_from_email(email_text: str) -> AWS_SNSMessageJSON:
     """
-    Assert the output equals the expected email, after replacements.
+    Create an SNS notification from a raw serialized email.
 
-    This function replaces MIME boundary strings, which change each test run.  The
-    replacement is "==[BOUNDARY#]==", where "#" is the order the string appears in the
-    email.
+    The notification will have the headers from the email, a passing receipt,
+    and other mocked items. The email will be included in the notification
+    body, not loaded from (mock) S3.
+    """
+    email = message_from_string(email_text, policy=policy.default)
+    topic_arn = "arn:aws:sns:us-east-1:168781634622:ses-inbound-grelay"
+    sns_message = {
+        "notificationType": "Received",
+        "mail": {
+            "timestamp": email["Date"].datetime.isoformat(),
+            "source": email["From"].addresses[0].addr_spec,
+            "messageId": email["Message-ID"],
+            "destination": [addr.addr_spec for addr in email["To"].addresses],
+            "headersTruncated": False,
+            "headers": [{"name": _h, "value": _v} for _h, _v in email.items()],
+            "commonHeaders": {
+                "from": [str(addr) for addr in email["From"].addresses],
+                "date": email["Date"],
+                "to": [str(addr) for addr in email["To"].addresses],
+                "messageId": email["Message-ID"],
+                "subject": email["Subject"],
+            },
+        },
+        "receipt": {
+            "timestamp": (email["Date"].datetime + timedelta(seconds=1)).isoformat(),
+            "processingTimeMillis": 1001,
+            "recipients": [addr.addr_spec for addr in email["To"].addresses],
+            "spamVerdict": {"status": "PASS"},
+            "virusVerdict": {"status": "PASS"},
+            "spfVerdict": {"status": "PASS"},
+            "dkimVerdict": {"status": "PASS"},
+            "dmarcVerdict": {"status": "PASS"},
+            "action": {"type": "SNS", "topicArn": topic_arn, "encoding": "UTF8"},
+        },
+        "content": email_text,
+    }
+    base_url = "https://sns.us-east-1.amazonaws.example.com"
+    sns_notification = {
+        "Type": "Notification",
+        "MessageId": str(uuid4()),
+        "TopicArn": topic_arn,
+        "Subject": email["Subject"],
+        "Message": json.dumps(sns_message),
+        "Timestamp": (email["Date"].datetime + timedelta(seconds=2)).isoformat(),
+        "SignatureVersion": "1",
+        "Signature": "invalid-signature",
```

Comment:
question (non-blocking): do we want the SNS to be hard-coded to an invalid signature?

Reply:
Yes, that's why it is hard-coded to the string "invalid-signature". It is supposed to be acceptable to `_sns_notification`, but not `sns_inbound`. I think it makes sense to clarify this in the docstring.

I think the Relay architecture for email processing is flawed, with the AWS provider details remaining a part of processing from start to finish. It restricts us to the one provider, and makes it harder to experiment with other systems in the future. If I had my preferences, I would not need to wrap emails in SNS structures at all, but could test them directly. We don't need that feature yet, so it will need to wait.